packages/autocomplete-core/src/utils/createCancelablePromise.ts

```diff
@@ -0,0 +1,160 @@
+import { noop } from '@algolia/autocomplete-shared';
+
+type CancelablePromiseState = {
+  isCanceled: boolean;
+  onCancelList: Array<(...args: any[]) => any>;
+};
+
+type PromiseExecutor<TValue> = (
+  resolve: (value: TValue | PromiseLike<TValue>) => void,
+  reject: (reason?: any) => void
+) => void;
+
+type CreateInternalCancelablePromiseParams<TValue> = {
+  executor?: PromiseExecutor<TValue>;
+  promise?: Promise<TValue>;
+  initialState?: CancelablePromiseState;
+};
+
+function createInternalCancelablePromise<TValue>({
+  executor = noop,
+  initialState = createInitialState(),
+  promise = new Promise<TValue>((resolve, reject) => {
+    return executor(resolve, reject);
+  }),
```

Comment:
as this is internal, the options could be required and passed in the calling code only. What do you think?

Reply:
What would be the benefit vs. having defaults?

Reply:
the output code of default values is slightly larger, and makes it slightly less clear what state lives "in" the function, and what is based on the declaration. It's not a big difference though I guess

Reply:
No you're right, it's actually nudging the numbers down and reduces the signature of `createInternalCancelablePromise` since we don't need to pass `executor` anymore.

Addressed in ffd16fe4.